commands/store.go

```diff
@@ -0,0 +1,240 @@
+	storeDeployCmd.Flags().BoolVar(&update, "update", false, "Update existing functions")
+	storeDeployCmd.Flags().StringArrayVar(&constraints, "constraint", []string{}, "Apply a constraint to the function")
+	storeDeployCmd.Flags().StringArrayVar(&secrets, "secret", []string{}, "Give the function access to a secure secret")
+
+	// Set bash-completion.
+	_ = storeDeployCmd.Flags().SetAnnotation("handler", cobra.BashCompSubdirsInDir, []string{})
+
+	storeCmd.AddCommand(storeListCmd)
+	storeCmd.AddCommand(storeInspectCmd)
+	storeCmd.AddCommand(storeDeployCmd)
+	faasCmd.AddCommand(storeCmd)
+}
+
+var storeCmd = &cobra.Command{
+	Use:   `store`,
+	Short: "OpenFaaS store commands",
+	Long:  "Allows browsing and deploying OpenFaaS store functions",
+}
+
+var storeListCmd = &cobra.Command{
+	Use:     `list [--store STORE_URL]`,
+	Short:   "List OpenFaaS store items",
+	Long:    "Lists the available items in OpenFaas store",
+	Example: `  faas-cli store list --store https://domain:port`,
+	RunE:    runStoreList,
+}
+
+var storeInspectCmd = &cobra.Command{
+	Use:     `inspect (FUNCTION_NAME|FUNCTION_TITLE) [--store STORE_URL]`,
+	Short:   "Show OpenFaaS store function details",
+	Long:    "Prints the detailed informations of the specified OpenFaaS function",
+	Example: `  faas-cli store inspect NodeInfo --store https://domain:port`,
+	RunE:    runStoreInspect,
+}
+
+var storeDeployCmd = &cobra.Command{
+	Use: `deploy (FUNCTION_NAME|FUNCTION_TITLE)
+							[--lang <ruby|python|node|csharp>]
+							[--gateway GATEWAY_URL]
+							[--handler HANDLER_DIR]
+							[--env ENVVAR=VALUE ...]
+							[--label LABEL=VALUE ...]
+							[--replace=false]
+							[--update=false]
+							[--constraint PLACEMENT_CONSTRAINT ...]
+							[--regex "REGEX"]
+							[--filter "WILDCARD"]
+							[--secret "SECRET_NAME"]`,
+
+	Short: "Deploy OpenFaaS functions from the store",
+	Long:  `Same as faas-cli deploy except pre-loaded with arguments from the store`,
+	Example: `  faas-cli store deploy figlet
+							faas-cli store deploy figlet
+									--gateway=http://remote-site.com:8080 --lang=python
+									--env=MYVAR=myval`,
+	RunE: runStoreDeploy,
+}
+
+func runStoreList(cmd *cobra.Command, args []string) error {
+	items, err := storeList(storeAddress)
+	if err != nil {
+		return err
+	}
+
+	if len(items) == 0 {
+		fmt.Printf("The store is empty.")
+		return nil
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
+	fmt.Fprintln(w)
+	fmt.Fprintln(w, "FUNCTION\tDESCRIPTION")
+
+	for _, item := range items {
+		fmt.Fprintf(w, "%s\t%s\n", item.Title, item.Description)
+	}
+
+	fmt.Fprintln(w)
+	w.Flush()
+
+	return nil
+}
+
+func runStoreInspect(cmd *cobra.Command, args []string) error {
+	if len(args) < 1 {
+		return fmt.Errorf("please provide the function name")
+	}
+
+	item, err := findFunction(args[0])
+	if err != nil {
+		return err
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
```

Comment:
Are you able to write this all into a string that is printed rather than directly to `os.Stdout`?

The `\t` mechanism may not do what you expect in all circumstances. I have used spacing instructions for `fmt.Sprintf()` before I think they would work well here.

Example: https://gobyexample.com/string-formatting

Reply:
That's how I started working on this but finally I found [`tabwriter`](https://golang.org/pkg/text/tabwriter/) which helps generating text-based tables. `\t` here is just a separator character it could be anything, it's not rendering tab character at all. Moreover, it won't render the text directly to output as it needs to receive all the rows data and calculate the required space for each column.